### PR TITLE
Added public API for listing all feature toggles

### PR DESCRIPTION
**Pull request by codebutler**

### CODE REVIEW

**Comment by jrbarron** [comment]
@codebutler Thanks for the contribution! I'll review the implementation shortly ;)

@ivarconr What do you think about listing the features as a general concept? Do the other clients support this already?

File: repository.go [hunk 1]
```diff
@@ -150,6 +150,14 @@ func (r *repository) getToggle(key string) *api.Feature {
 	return nil
 }
 
+func (r *repository) list() []api.Feature {
+	var features []api.Feature
```
Comment (jrbarron) [inline repository.go]: You should probably lock the mutex for reading here before reading from the store. This should just be a matter of adding:

```go
r.RLock()
defer r.RUnlock()
```
to the top of the function.

File: feature/feature.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package feature
+
+type Feature struct {
```
Comment (jrbarron) [inline feature/feature.go]: The struct itself and the fields should be documented since they are exported.

**Comment by ivarconr** [comment]
> @ivarconr What do you think about listing the features as a general concept? Do the other clients support this already?

Yes, we have seen that it from time to time are valid use cases where one would like to get the full feature toggle definitions. In the node client you can use the `getFeatureToggleDefinitions()`:

https://github.com/Unleash/unleash-client-node#3-use-unleash

File: client.go [hunk 3]
```diff
@@ -347,3 +348,16 @@ func (uc *Client) getStrategy(name string) strategy.Strategy {
...
+
+// ListFeatures returns all available features toggles.
+func (uc *Client) ListFeatures() []*feature.Feature {
+	var features []*feature.Feature
+	for _, f := range uc.repository.list() {
+		features = append(features, &feature.Feature{
+			Name:         f.Name,
+			Description:  f.Description,
+			Enabled:      f.Enabled,
+		})
```
Comment (ivarconr) [inline client.go]: should also include all details (strategy and variant details missing)

Comment (codebutler) [reply client.go]: Would you like *all* fields to be copied over?

My intuition was to not expose the `api` data structures but this is a lot of duplication now.

**Comment by codebutler** [comment]
Please let me know if these updates look right and I'll add unit tests. Thanks!

**Comment by codebutler** [comment]
@jrbarron @ivarconr could you please take another look at this? ty!

**Comment by jrbarron** [comment]
@codebutler First of all, my apologies for not getting back to this sooner. I've looked at your latest changes and they look correct from a purely technical perspective. I'm struggling a bit with the amount of duplication though and I'm sure you felt the same way when implementing it :) 

Originally the idea was to keep the API surface area of the package as small as possible which is why the API structures were never exported. However I feel the introduction of your `ListFeatures` function necessitates making those structures publicly available instead of duplicating them. The only reason I can think of for duplicating them would be if we wanted to support the internal API structures being vastly different from the public facing ones and I don't really see this happening and even if it did, it would probably warrant a new major version of the package.

So if you agree with that sentiment (?) then the next thing to decide is where to put the files. I guess there is no point in continuing to have an `internal` directory if most things are exported. Is this something you would feel comfortable trying or would you like me to come with a concrete suggestion here?

**Comment by codebutler** [comment]
@ivanklee86 please take another look.

**Comment by codebutler** [comment]
@jrbarron ping

**Comment by codebutler** [comment]
Thanks! Who has permission to merge this in?

**Comment by ivarconr** [comment]
> Thanks! Who has permission to merge this in?

@jrbarron should merge this. (He might be on holiday, it's that season.)

**Comment by jrbarron** [comment]
Merged now (sorry, first day back from vacation)

**Comment by dcelasun** [comment]
Hey guys, can you please tag a new release with this commit? It makes dependency management easier.

**Pull request merged**